shielded-pool/src/state_key.rs

```diff
@@ -0,0 +1,55 @@
+use jmt::KeyHash;
+use penumbra_chain::CompactBlock;
+use penumbra_crypto::{
+    asset::{Denom, Id as Asset_Id},
+    note, Nullifier,
+};
+use penumbra_tct::Root;
+
+pub fn token_supply(asset_id: &Asset_Id) -> KeyHash {
+    format!("shielded_pool/assets/{}/token_supply", asset_id).into()
+}
+
+pub fn known_assets() -> KeyHash {
+    format!("shielded_pool/known_assets").into()
+}
+
+pub fn denom_by_asset(asset_id: &Asset_Id) -> KeyHash {
+    format!("shielded_pool/assets/{}/denom", asset_id).into()
+}
+
+pub fn register_denom(denom: &Denom) -> KeyHash {
+    format!("shielded_pool/assets/{}/denom", denom).into()
+}
```

Comment:
These are the same function; the `register_denom` function should be removed, and replaced by calls to `denom_by_asset`.

That way, we just have one function _per key_, regardless of the context in which that key is used.

Reply:
can just use generic right? Where T can be denom asset_id or denom??